From 69fe9f0947b34a20213aa7fa107d63352c7f2d18 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Mon, 14 Jul 2014 13:07:59 -0400
Subject: [scsi] bnx2fc: Fix CPU hotplug callback registration

Message-id: <1405343292-21004-23-git-send-email-prarit@redhat.com>
Patchwork-id: 85772
O-Subject: [RHEL7.1 PATCH BZ 1119078 22/35] scsi, bnx2fc: Fix CPU hotplug callback registration
Bugzilla: 1119078
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: Tomas Henzl <thenzl@redhat.com>
RH-Acked-by: Igor Mammedov <imammedo@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1119078

commit 7229b6d0b2a9e62d5eb4886dcd0137664c735718
Author: Srivatsa S. Bhat <srivatsa.bhat@linux.vnet.ibm.com>
Date:   Tue Mar 11 02:09:45 2014 +0530

    scsi, bnx2fc: Fix CPU hotplug callback registration

    Subsystems that want to register CPU hotplug callbacks, as well as perform
    initialization for the CPUs that are already online, often do it as shown
    below:

     get_online_cpus();

     for_each_online_cpu(cpu)
      init_cpu(cpu);

     register_cpu_notifier(&foobar_cpu_notifier);

     put_online_cpus();

    This is wrong, since it is prone to ABBA deadlocks involving the
    cpu_add_remove_lock and the cpu_hotplug.lock (when running concurrently
    with CPU hotplug operations).

    Instead, the correct and race-free way of performing the callback
    registration is:

     cpu_notifier_register_begin();

     for_each_online_cpu(cpu)
      init_cpu(cpu);

     /* Note the use of the double underscored version of the API */
     __register_cpu_notifier(&foobar_cpu_notifier);

     cpu_notifier_register_done();

    Fix the bnx2fc code in scsi by using this latter form of callback
    registration.

    Cc: Eddie Wai <eddie.wai@broadcom.com>
    Cc: "James E.J. Bottomley" <JBottomley@parallels.com>
    Cc: Ingo Molnar <mingo@kernel.org>
    Signed-off-by: Srivatsa S. Bhat <srivatsa.bhat@linux.vnet.ibm.com>
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Maurizio Lombardi <mlombard@redhat.com>
Cc: David Milburn <dmilburn@redhat.com>
Cc: Doug Ledford <dledford@redhat.com>
Cc: Tomas Henzl <thenzl@redhat.com>
Cc: Tony Camuso <tcamuso@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/bnx2fc/bnx2fc_fcoe.c b/drivers/scsi/bnx2fc/bnx2fc_fcoe.c
index 7319db6..0c03f5e 100644
--- a/drivers/scsi/bnx2fc/bnx2fc_fcoe.c
+++ b/drivers/scsi/bnx2fc/bnx2fc_fcoe.c
@@ -2593,12 +2593,16 @@ static int __init bnx2fc_mod_init(void)
   spin_lock_init(&p->fp_work_lock);
  }
 
+ cpu_notifier_register_begin();
+
  for_each_online_cpu(cpu) {
   bnx2fc_percpu_thread_create(cpu);
  }
 
  /* Initialize per CPU interrupt thread */
- register_hotcpu_notifier(&bnx2fc_cpu_notifier);
+ __register_hotcpu_notifier(&bnx2fc_cpu_notifier);
+
+ cpu_notifier_register_done();
 
  cnic_register_driver(CNIC_ULP_FCOE, &bnx2fc_cnic_cb);
 
@@ -2663,13 +2667,17 @@ static void __exit bnx2fc_mod_exit(void)
  if (l2_thread)
   kthread_stop(l2_thread);
 
- unregister_hotcpu_notifier(&bnx2fc_cpu_notifier);
+ cpu_notifier_register_begin();
 
  /* Destroy per cpu threads */
  for_each_online_cpu(cpu) {
   bnx2fc_percpu_thread_destroy(cpu);
  }
 
+ __unregister_hotcpu_notifier(&bnx2fc_cpu_notifier);
+
+ cpu_notifier_register_done();
+
  destroy_workqueue(bnx2fc_wq);
  /*
   * detach from scsi transport
-- 
1.7.1